components/VM_Arm/src/main.c

```diff
@@ -1174,6 +1184,16 @@ int main_continued(void)
         return -1;
     }
 
+    /* Install and initialize guest RAM. There might be no memory here, because
+     * this is the exclusive fixed guest memory. Modules may set up other forms
+     * of guest memory, e.g. memory shared with other VM and memory that is
+     * mapped on demand.
+     */
+    if (ram_size > 0) {
```

Comment:
if ram_size is 0 or less than zero, we have a problem. I'd change this to:

```
ZF_LOGF_IF(0 > ram_size, "ram_size must be a positive value, as running a guest without RAM is impossible");

err = vm_ram_register_at(&vm, ram_base, ram_size, vm.mem.map_one_to_one);
assert(!err);
```

Reply:
`ram_size` is an `unsigned long`, so this can never be less than zero. With https://github.com/seL4/camkes-vm/pull/63 it will become `size_t`, which also can't ever get negative. We have zero to disable it now. Allowing negative values here has no use case and it seem to just create more trouble.

Reply:
Fair point about the negative, but this would still mean a value of 0 is possible, when it would cause a host of issues.

Reply:
> The changes look good to me.
> 
> Is your logic for removing the init_ram module just that its a small function? In my mind, modules should be for optional things. So things like device tree generation, virtio devices, etc. Basically you want to end up with as small of a VMM as possible, right? Since RAM initialization isn't optional (you need it to load in the guest images), I'm okay with removing it as a module.

The first part I agree. For example, we have need to customize FDT. I was thinking perhaps we could have `DEFINE_MODULE_EXT` which allows us to specify hooks, for example, during the device tree generation phase. The second that comes to my mind that the fault handler should be extensible with modules. Of course using the MMIO mapping functions is an option, but not sure if we need hooks for `unhandled_mem_fault_callback()` (that's how we have been doing it, but not sure if it's 100% necessary).

I was going to say to I disagree with the second part, but but... @axel-h  how come could we run our "map guest RAM from dataport" before image loading is done? Perhaps hook or weak function is needed after all?

Reply:
Still some thoughts about modules... I think in more complex systems they will have dependencies so that hooks from module B must be ran before ones from module A. This extended version of `DEFINE_MODULE` macro should handle those as well. Perhaps it needs a little more thinking and solving this with weak function is best for short term solution.

Reply:
> I was going to say to I disagree with the second part, but but... @axel-h how come could we run our "map guest RAM 
> from dataport" before image loading is done? Perhaps hook or weak function is needed after all?

@hlyytine Not sure I get the question. Modules are initialized before images are loaded, so they can provide memory that the modules are put into then. This will not change with this PR. It's just that this RAM init now runs before any module runs. So I don't think a weak hook needs to be provided.

Reply:
> > I was going to say to I disagree with the second part, but but... @axel-h how come could we run our "map guest RAM
> > from dataport" before image loading is done? Perhaps hook or weak function is needed after all?
> 
> @hlyytine Not sure I get the question. Modules are initialized before images are loaded, so they can provide memory that the modules are put into then. This will not change with this PR. It's just that this RAM init now runs before any module runs. So I don't think a weak hook needs to be provided.

OK, that seems to be case. No need for weak function for that reason and I'm fine with that (provided that you give us a way to generate memory node to DTB). But how about backwards compatibility? That init_ram_module() function has been declared weak since ARM VMM got merged into camkes-vm repo and I wouldn't be surprised if somebody somewhere has overridden it. Shall we summon @kent-mcleod to give his opinion?

Reply:
> * the memory node is not created at all in the DTB. I consider this more sane than creating a node with a zero size. But in the end I'm collecting input here and would add this in a comment then why we need such a node in any case.

In our use case (map guest RAM from dataport) we need memory node in DTB. We are using autogenerated DTB, so just statically providing it is not enough for us.

Reply:
> In our use case (map guest RAM from dataport) we need memory node in DTB. We are using autogenerated DTB, so just statically providing it is not enough for us.

@hlyytine With the changes in this PR, no "ordinary" memory node is created. Your module would have to create one for the memory it adds. However, this seem impossible in the current code, but I have another PR coming that that allows functions and modules to manipulate the generated DTB according to their "contributions" to the VM. Seems this is exactly what you need then.